### PR TITLE
common: use mono clock for HeartbeatMap

### PR DESCRIPTION
**Pull request by XinzeChi**
Signed-off-by: Xinze Chi <xinze@xsky.com>

### CODE REVIEW

**Comment by xiexingguo** [comment]
s/heaerbeatmap/HeartbeatMap/

File: src/common/HeartbeatMap.h [hunk 1]
```diff
@@ -79,12 +80,12 @@ class HeartbeatMap {
  private:
   CephContext *m_cct;
   RWLock m_rwlock;
-  time_t m_inject_unhealthy_until;
+  unsigned m_inject_unhealthy_until;
```
Comment (liewegas) [inline src/common/HeartbeatMap.h]: if this is a time_point instead wouldn't that avoid most of the casts?

File: src/common/HeartbeatMap.h [hunk 2]
```diff
@@ -79,12 +80,12 @@ class HeartbeatMap {
...
   std::list<heartbeat_handle_d*> m_workers;
   std::atomic<unsigned> m_unhealthy_workers = { 0 };
   std::atomic<unsigned> m_total_workers = { 0 };
 
-  bool _check(const heartbeat_handle_d *h, const char *who, time_t now);
+  bool _check(const heartbeat_handle_d *h, const char *who, ceph::coarse_mono_clock::time_point& now);
```
Comment (tchaikov) [inline src/common/HeartbeatMap.h]: pass by `const ceph::coarse_mono_clock::time_point&`

Comment (amitkumar50) [reply src/common/HeartbeatMap.h]: Agree with kefu.

File: src/common/HeartbeatMap.cc [hunk 3]
```diff
@@ -64,19 +64,19 @@ void HeartbeatMap::remove_worker(const heartbeat_handle_d *h)
...
-bool HeartbeatMap::_check(const heartbeat_handle_d *h, const char *who, time_t now)
+bool HeartbeatMap::_check(const heartbeat_handle_d *h, const char *who, ceph::coarse_mono_clock::time_point &now)
 {
   bool healthy = true;
-  time_t was;
+  unsigned was;
 
   was = h->timeout;
-  if (was && was < now) {
+  if (was && was < std::chrono::time_point_cast<std::chrono::seconds>(now).time_since_epoch().count()) {
```
Comment (amitkumar50) [inline src/common/HeartbeatMap.cc]: would be good to use typdef for 'std::chrono::time_point_cast<std::chrono::seconds>' instead repeating again & again

**Comment by tchaikov** [comment]
@XinzeChi ping.

**Comment by XinzeChi** [comment]
@tchaikov pong

File: src/common/HeartbeatMap.h (hunk 1 above)
Comment (tchaikov) [reply src/common/HeartbeatMap.h]: agree, would be easier if this is a `time_point`.

File: src/common/HeartbeatMap.cc [hunk 4]
```diff
@@ -90,14 +96,14 @@ void HeartbeatMap::reset_timeout(heartbeat_handle_d *h, time_t grace, time_t sui
 {
   ldout(m_cct, 20) << "reset_timeout '" << h->name << "' grace " << grace
 		   << " suicide " << suicide_grace << dendl;
-  time_t now = time(NULL);
+  auto now = ceph::coarse_mono_clock::now();
   _check(h, "reset_timeout", now);
 
-  h->timeout = now + grace;
+  h->timeout = tp_to_seconds(now) + grace;
```
Comment (tchaikov) [inline src/common/HeartbeatMap.cc]: might want to use `std::atomic<ceph::coarse_mono_clock>` for `heartbeat_handle_d::timeout` and `heartbeat_handle_d::suicide_timeout`

Comment (XinzeChi) [reply src/common/HeartbeatMap.cc]: it looks complicated if we set std::atomic<ceph::coarse_mono_clock>. https://stackoverflow.com/questions/29364456/stdatomicstdchronohigh-resolution-clocktime-point-can-not-compile

Comment (tchaikov) [reply src/common/HeartbeatMap.cc]: #17213

**Comment by tchaikov** [comment]
closing in favor of #17213

File: src/common/HeartbeatMap.cc (hunk 4 above)
Comment (tchaikov) [reply src/common/HeartbeatMap.cc]: @XinzeChi sorry, i just figured out that this is a bug in GCC. see https://gcc.gnu.org/bugzilla/show_bug.cgi?id=53901 and https://gcc.gnu.org/bugzilla/show_bug.cgi?id=53903.

**Pull request closed without merging**